### PR TITLE
Added sublist support for torch.einsum

### PR DESCRIPTION
**Pull request by heitorschueroff**
This PR adds an alternative way of calling `torch.einsum`. Instead of specifying the subscripts as letters in the `equation` parameter, one can now specify the subscripts as a list of integers as in `torch.einsum(operand1, subscripts1, operand2, subscripts2, ..., [subscripts_out])`. This would be equivalent to `torch.einsum('<subscripts1>,<subscripts2>,...,->[<subscript_out>]', operand1, operand2, ...)`

TODO
- [x] Update documentation
- [x] Add more error checking
- [x] Update tests

### CODE REVIEW

File: test/test_linalg.py
Comment (mruberry) [inline test/test_linalg.py:4466]: nit: use make_tensor()

File: torch/functional.py
Comment (mruberry) [inline torch/functional.py:352]: What do you think about separating the docs update so we can get higher code velocity and focus on the docs in a follow-up PR?

For now, for example, this note could read, "As of PyTorch 1.9 :func:\`torch.einsum\` also supports the sublist format (see examples below)." ?

File: torch/functional.py
Comment (mruberry) [inline torch/functional.py:361]: Nice type fix

File: torch/functional.py
Comment (mruberry) [inline torch/functional.py:363]: Add a sublist example

**Review by mruberry — COMMENTED** [review]
Hey @heitorschueroff!

The approach this PR takes and its consistency with the existing einsum work makes a lot of sense. I made a few suggestions about how to handle the docs in this PR, mostly with an eye towards code velocity. Otherwise I think it's nearly good to go! Looking forward to hearing your thoughts.

File: test/test_linalg.py
Comment (mruberry) [inline test/test_linalg.py:4459]: Should we add a jit sublist test?

File: torch/functional.py
Comment (heitorschueroff) [reply torch/functional.py:352]: Sure, I'll do that.

File: torch/functional.py
Comment (heitorschueroff) [reply torch/functional.py:363]: Will do in a follow up PR

File: test/test_linalg.py
Comment (heitorschueroff) [reply test/test_linalg.py:4459]: The sublist format cannot be scripted because the argument types do not match the C++ signature. Is there a way to specify operations that do not support jit?

File: torch/functional.py
Comment (heitorschueroff) [reply torch/functional.py]: PR adding doc here https://github.com/pytorch/pytorch/pull/57038

**Comment by heitorschueroff** [comment]
@mruberry I separated the doc update here https://github.com/pytorch/pytorch/pull/57038. For the JIT test, sublist format cannot be scripted because the input types don't match the C++ signature, it can be traced however. How should I handle this?

**Comment by mruberry** [comment]
> @mruberry I separated the doc update here #57038. For the JIT test, sublist format cannot be scripted because the input types don't match the C++ signature, it can be traced however. How should I handle this?

What about adding a test that it can be traced, and a test that attempting to script works for non-sublist inputs, but fails on sublist inputs?

Like 1 non-sublist example and 1 sublist example, and then the cross-product of those examples with scripting and tracing?

**Comment by heitorschueroff** [comment]
> > @mruberry I separated the doc update here #57038. For the JIT test, sublist format cannot be scripted because the input types don't match the C++ signature, it can be traced however. How should I handle this?
> 
> What about adding a test that it can be traced, and a test that attempting to script works for non-sublist inputs, but fails on sublist inputs?
> 
> Like 1 non-sublist example and 1 sublist example, and then the cross-product of those examples with scripting and tracing?

Yeah, that's what I had but I was wondering if there was a more formal way of specifying unsupported features in JIT.

**Comment by mruberry** [comment]
> Yeah, that's what I had in mind but I was wondering if there was a more formal way of specifying unsupported features in JIT.

I don't think there is, but it might be interesting to refine our OpInfo jit tests/metadata

**Comment by heitorschueroff** [comment]
> > @mruberry I separated the doc update here #57038. For the JIT test, sublist format cannot be scripted because the input types don't match the C++ signature, it can be traced however. How should I handle this?
> 
> What about adding a test that it can be traced, and a test that attempting to script works for non-sublist inputs, but fails on sublist inputs?
> 
> Like 1 non-sublist example and 1 sublist example, and then the cross-product of those examples with scripting and tracing?

Done

File: aten/src/ATen/native/Linear.cpp
Comment (mruberry) [inline aten/src/ATen/native/Linear.cpp:155]: Nice explicit typing

File: torch/functional.py
Comment (mruberry) [inline torch/functional.py:408]: Nit: comment explaining the format and how the conversion occurs

**Review by mruberry — APPROVED** [review]
This looks really good, I made a small suggestion

think we should wait until after the branch cut to merge this to give it bake time?

**Pull request closed without merging**
